packages/field-base/test/text-field-mixin.test.js

```diff
@@ -0,0 +1,201 @@
+    });
+
+    it('should propagate maxlength property to the input', () => {
+      element.maxlength = 8;
+      expect(input.maxLength).to.equal(8);
+    });
+
+    it('should propagate minlength property to the input', () => {
+      element.minlength = 8;
+      expect(input.minLength).to.equal(8);
+    });
+  });
+
+  describe('validation', () => {
+    it('should not change invalid property if no constraints are set', () => {
+      element.validate();
+      expect(element.invalid).to.be.false;
+      element.invalid = true;
+      element.validate();
+      expect(element.invalid).to.be.true;
+    });
+
+    it('should validate the field when invalid and required is set', () => {
+      element.invalid = true;
+      const spy = sinon.spy(element, 'validate');
+      element.required = true;
+      expect(spy.calledOnce).to.be.true;
+    });
+
+    it('should validate the field when invalid after minlength is changed', () => {
+      element.invalid = true;
+      const spy = sinon.spy(element, 'validate');
+      element.minlength = 2;
+      expect(spy.calledOnce).to.be.true;
+    });
+
+    it('should validate the field when invalid after maxlength is changed', () => {
+      element.invalid = true;
+      const spy = sinon.spy(element, 'validate');
+      element.maxlength = 6;
+      expect(spy.calledOnce).to.be.true;
+    });
+
+    it('should validate the field when invalid after pattern is changed', () => {
+      element.invalid = true;
+      const spy = sinon.spy(element, 'validate');
+      element.pattern = '[-+\\d]';
+      expect(spy.calledOnce).to.be.true;
+    });
+
+    it('should override explicitly set invalid if constraints are set', () => {
+      element.invalid = true;
+      element.value = 'foo';
+      element.required = true;
+      expect(element.invalid).to.be.false;
+    });
+
+    it('should update invalid state when required is removed', () => {
+      element.required = true;
+      element.validate();
+      expect(element.invalid).to.be.true;
+
+      element.required = false;
+      expect(element.invalid).to.be.false;
+    });
+
+    it('should update invalid state when pattern is removed', () => {
+      element.value = '123foo';
+      element.pattern = '\\d+';
+      element.validate();
+      expect(element.invalid).to.be.true;
+
+      element.pattern = '';
+      expect(element.invalid).to.be.false;
+    });
+  });
+
+  describe('prevent invalid input', () => {
+    beforeEach(() => {
+      element.preventInvalidInput = true;
+      element.value = '1';
+    });
+
+    describe('user action', () => {
+      function inputText(value) {
+        input.value = value;
+        input.dispatchEvent(new CustomEvent('input'));
+      }
+
+      it('should prevent invalid pattern', () => {
+        element.pattern = '[0-9]*';
+        inputText('f');
+        expect(element.value).to.equal('1');
+      });
+
+      it('should temporarily set input-prevented attribute on invalid input', () => {
```

Comment:
Could also add a test that input-prevented is removed after a delay

Reply:
Done.